File: core-csharp/src/MathCursor.Core/Parsing/Parser.cs
using System.Collections.Generic;
using MathCursor.Core.Ast;

namespace MathCursor.Core.Parsing;

/// <summary>
/// Parser recursive-descent des expressions math.
/// Précédence : ^ > */ > +-=,
/// Porté depuis archive/officejs-prototype/src/taskpane/conversion/parser.ts.
/// </summary>
public static class Parser
{
    private sealed class State { public int I; }

    public static MathNode Parse(IList<LexToken> tokens)
    {
        var state = new State { I = 0 };
        return ParseAdditive(tokens, state);
    }

    // additive : [+|-|=] term ((+|-|=|,) term)*
    private static MathNode ParseAdditive(IList<LexToken> tks, State p)
    {
        MathNode left;
        if (p.I < tks.Count && tks[p.I].Kind == LexTokenKind.Op && "+-=".Contains(tks[p.I].Value))
        {
            var op = tks[p.I++].Value;
            left = new UnaryNode { Op = op, Child = ParseMultiplicative(tks, p) };
        }
        else
        {
            left = ParseMultiplicative(tks, p);
        }

        while (p.I < tks.Count && tks[p.I].Kind == LexTokenKind.Op && "+-=,".Contains(tks[p.I].Value))
        {
            var op = tks[p.I++].Value;
            left = new BinaryOpNode { Op = op, Left = left, Right = ParseMultiplicative(tks, p) };
        }
        return left;
    }

    // multiplicative : power ((* | / | juxtaposition) power)*
    private static MathNode ParseMultiplicative(IList<LexToken> tks, State p)
    {
        var left = ParsePower(tks, p);
        while (p.I < tks.Count)
        {
            var nx = tks[p.I];
            if (nx.Kind == LexTokenKind.Op && nx.Value == "/")
            {
                p.I++;
                left = new FractionNode { Numerator = left, Denominator = ParsePower(tks, p) };
                continue;
            }
            if (nx.Kind == LexTokenKind.Op && nx.Value == "*")
            {
                p.I++;
                left = new BinaryOpNode { Op = "\u00D7", Left = left, Right = ParsePower(tks, p) };
                continue;
            }
            if (nx.Kind == LexTokenKind.Number || nx.Kind == LexTokenKind.Variable
                || nx.Kind == LexTokenKind.LParen || nx.Kind == LexTokenKind.LBracket)
            {
                var right = ParsePower(tks, p);
                if (left is JuxtapositionNode j) { j.Parts.Add(right); }
                else { left = new JuxtapositionNode { Parts = new List<MathNode> { left, right } }; }
                continue;
            }
            break;
        }
        return left;
    }

    // power : unary ^ power (droite-associatif)
    private static MathNode ParsePower(IList<LexToken> tks, State p)
    {
        if (p.I < tks.Count && tks[p.I].Kind == LexTokenKind.Op && "+-".Contains(tks[p.I].Value))
        {
            var op = tks[p.I++].Value;
            return new UnaryNode { Op = op, Child = ParsePower(tks, p) };
        }
        var basis = ParseAtom(tks, p);
        if (p.I < tks.Count && tks[p.I].Kind == LexTokenKind.Op && tks[p.I].Value == "^")
        {
            p.I++;
            return new SuperscriptNode { Base = basis, Exponent = ParsePower(tks, p) };
        }
        return basis;
    }

    // atom : nombre | variable | (expr) | [expr]
    private static MathNode ParseAtom(IList<LexToken> tks, State p)
    {
        if (p.I >= tks.Count) return new EmptyNode();
        var tk = tks[p.I];

        if (tk.Kind == LexTokenKind.LParen || tk.Kind == LexTokenKind.LBracket)
        {
            var open = tk.Kind == LexTokenKind.LParen ? "(" : "[";
            var close = tk.Kind == LexTokenKind.LParen ? ")" : "]";
            p.I++;
            var inner = ParseAdditive(tks, p);
            if (p.I < tks.Count && tks[p.I].Value == close) p.I++;
            return new ParenNode { OpenChar = open, Inner = inner };
        }

        if (tk.Kind == LexTokenKind.Number) { p.I++; return new NumberNode { Value = tk.Value }; }
        if (tk.Kind == LexTokenKind.Variable) { p.I++; return new VariableNode { Name = tk.Value }; }

        p.I++;
        return new VariableNode { Name = tk.Value };
    }
}

File: src/taskpane/conversion/parser.ts
// Phase 2 : Tokens → AST
// Précédence: ^ > */ > +-=,

import type { N, Tk } from "./types";

export function parse(tks: Tk[]): N {
  const p = { i: 0 };
  return pAdd(tks, p);
}

// additive: [=|+|-] term ((+|-|=|,) term)*
function pAdd(tks: Tk[], p: { i: number }): N {
  let left: N;
  if (p.i < tks.length && tks[p.i].t === "op" && "+-=".includes(tks[p.i].v)) {
    const op = tks[p.i++].v;
    left = { k: "unary", op, child: pMul(tks, p) };
  } else {
    left = pMul(tks, p);
  }
  while (p.i < tks.length && tks[p.i].t === "op" && "+-=,".includes(tks[p.i].v)) {
    const op = tks[p.i++].v;
    left = { k: "op", op, left, right: pMul(tks, p) };
  }
  return left;
}

// multiplicative: power ((* | / | juxtaposition) power)*
function pMul(tks: Tk[], p: { i: number }): N {
  let left = pPow(tks, p);
  while (p.i < tks.length) {
    const nx = tks[p.i];
    if (nx.t === "op" && nx.v === "/") {
      p.i++;
      left = { k: "frac", num: left, den: pPow(tks, p) };
      continue;
    }
    if (nx.t === "op" && nx.v === "*") {
      p.i++;
      left = { k: "op", op: "\u00D7", left, right: pPow(tks, p) };
      continue;
    }
    if (nx.t === "n" || nx.t === "v" || nx.t === "(" || nx.t === "[") {
      const right = pPow(tks, p);
      if (left.k === "juxt") { left.parts.push(right); }
      else { left = { k: "juxt", parts: [left, right] }; }
      continue;
    }
    break;
  }
  return left;
}

// power: unary ^ power (droite-associatif)
function pPow(tks: Tk[], p: { i: number }): N {
  if (p.i < tks.length && tks[p.i].t === "op" && "+-".includes(tks[p.i].v)) {
    const op = tks[p.i++].v;
    return { k: "unary", op, child: pPow(tks, p) };
  }
  const base = pAtom(tks, p);
  if (p.i < tks.length && tks[p.i].t === "op" && tks[p.i].v === "^") {
    p.i++;
    return { k: "sup", base, exp: pPow(tks, p) };
  }
  return base;
}

// atom: nombre | variable | (expr) | [expr]
function pAtom(tks: Tk[], p: { i: number }): N {
  if (p.i >= tks.length) return { k: "empty" };
  const tk = tks[p.i];
  if (tk.t === "(" || tk.t === "[") {
    const d = tk.t as "(" | "[";
    const close = d === "(" ? ")" : "]";
    p.i++;
    const inner = pAdd(tks, p);
    if (p.i < tks.length && tks[p.i].v === close) p.i++;
    return { k: "paren", d, inner };
  }
  if (tk.t === "n") { p.i++; return { k: "num", v: tk.v }; }
  if (tk.t === "v") { p.i++; return { k: "var", v: tk.v }; }
  p.i++;
  return { k: "var", v: tk.v };
}

// Debug: AST → string
export function astToString(n: N, depth = 0): string {
  const pad = "  ".repeat(depth);
  switch (n.k) {
    case "empty": return `${pad}(empty)`;
    case "num": return `${pad}num(${n.v})`;
    case "var": return `${pad}var(${n.v})`;
    case "op": return `${pad}op(${n.op})\n${astToString(n.left, depth + 1)}\n${astToString(n.right, depth + 1)}`;
    case "unary": return `${pad}unary(${n.op})\n${astToString(n.child, depth + 1)}`;
    case "frac": return `${pad}frac\n${pad}  num:\n${astToString(n.num, depth + 2)}\n${pad}  den:\n${astToString(n.den, depth + 2)}`;
    case "sup": return `${pad}sup\n${astToString(n.base, depth + 1)}\n${astToString(n.exp, depth + 1)}`;
    case "paren": return `${pad}paren(${n.d})\n${astToString(n.inner, depth + 1)}`;
    case "juxt": return `${pad}juxt\n${n.parts.map(p => astToString(p, depth + 1)).join("\n")}`;
  }
}
